Userland/Libraries/LibWeb/WebAudio/AudioContext.cpp

```diff
@@ -31,4 +89,210 @@ JS::ThrowCompletionOr<void> AudioContext::initialize(JS::Realm& realm)
+    set_control_state(Bindings::AudioContextState::Suspended);
+
+    // 7. Queue a control message to suspend the AudioContext.
+    // FIXME: Implement control message queue to run following steps on the rendering thread
+
+    // FIXME: 7.1: Attempt to release system resources.
+
+    // 7.2: Set the [[rendering thread state]] on the AudioContext to suspended.
+    set_rendering_state(Bindings::AudioContextState::Suspended);
+
+    // 7.3: queue a media element task to execute the following steps:
+    queue_a_media_element_task([&]() {
+        // 7.3.1: Resolve promise.
+        *promise->resolve();
+
+        // 7.3.2: If the state attribute of the AudioContext is not already "suspended":
+        if (state() != Bindings::AudioContextState::Suspended) {
+            // 7.3.2.1: Set the state attribute of the AudioContext to "suspended".
+            set_control_state(Bindings::AudioContextState::Suspended);
+
+            // 7.3.2.2: queue a media element task to fire an event named statechange at the AudioContext.
+            queue_a_media_element_task([&]() {
+                this->dispatch_event(DOM::Event::create(realm, HTML::EventNames::statechange).release_value_but_fixme_should_propagate_errors());
+            });
+        }
+    });
+
+    // 8. Return promise.
+    return JS::NonnullGCPtr { verify_cast<JS::Promise>(*promise->promise()) };
+}
+
+// https://www.w3.org/TR/webaudio/#dom-audiocontext-close
+WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::Promise>> AudioContext::close()
+{
+    auto& realm = this->realm();
+
+    // FIXME: 1. If this's relevant global object's associated Document is not fully active then return a promise rejected with "InvalidStateError" DOMException.
+
+    // 2. Let promise be a new Promise.
+    auto promise = WebIDL::create_promise(realm);
+
+    // 3. If the [[control thread state]] flag on the AudioContext is closed reject the promise with InvalidStateError, abort these steps, returning promise.
+    if (state() == Bindings::AudioContextState::Closed) {
+        WebIDL::reject_promise(realm, promise, WebIDL::InvalidStateError::create(realm, "Audio context is already closed."));
+        return JS::NonnullGCPtr { verify_cast<JS::Promise>(*promise->promise()) };
+    }
+
+    // 4. Set the [[control thread state]] flag on the AudioContext to closed.
+    set_control_state(Bindings::AudioContextState::Closed);
+
+    // 5. Queue a control message to close the AudioContext.
+    // FIXME: Implement control message queue to run following steps on the rendering thread
+
+    // FIXME: 5.1: Attempt to release system resources.
+
+    // 5.2: Set the [[rendering thread state]] to "suspended".
+    set_rendering_state(Bindings::AudioContextState::Suspended);
+
+    // FIXME: 5.3: If this control message is being run in a reaction to the document being unloaded, abort this algorithm.
+
+    // 5.4: queue a media element task to execute the following steps:
+    queue_a_media_element_task([&]() {
```

Comment:
Please use a more explicit capture list than `[&]` since it's very easy to accidentally capture a stack local which then becomes invalid by the time the task is executed :)

Reply:
(Same comment for every time `[&]` is used)